### ISSUE TITLE
🐕 Batch: Unit Testing in Ersilia CLI

### ISSUE THREAD

**Issue by DhanshreeA** [ersilia-os/ersilia#1319]
### Summary

We need to implement unit tests for the different components within Ersilia to make sure our pipelines are robust and we are able to push with confidence. 

Due to the [work](https://github.com/ersilia-os/ersilia/pull/1295) of our fantastic contributors, we are expanding functionality within Ersilia that would definitely benefit from testing (PR linked as an example). This issue has a list of different components we could easily set up testing for. Please only pick one item, ie one checkbox` to work on and link your PR in this issue so we can track the ongoing work. 

### Objective(s)

1. Tests for `CompoundIdentifier` class in ersilia/utils/identifiers/compound.py. Particularly, we need the following tests, and we will mock response from all the resolver URLs used in this class:
- [ ] `is_input_header` positive test 
- [ ] `is_input_header` negative test
- [ ] `is_key_header` positive test
- [ ] `is_key_header` negative test
- [ ] `_is_smiles` positive test when Chem is None
- [ ] `_is_smiles` positive test when Chem is not None
- [ ] `_is_smiles` negative test when Chem is None
- [ ] `_is_smiles` negative test when Chem is not None
- [ ] `_is_inchikey` positive test
- [ ] `_is_inchikey` negative test
- [ ] `guess_type` with inchikey
- [ ] `guess_type` with smiles
- [ ] `guess_type` with UNPROCESSABLE_INPUT
- [ ] `guess_type` with whitespaces (\n, \t, etc) as input
- [ ] `guess_type` with non character input (numbers, floats, random unicode characters)
- [ ] `_nci_smiles_to_inchikey` positive test (we will mock this URL)
- [ ] `_nci_smiles_to_inchikey` negative test (mock this URL again)
- [ ] `_pubchem_smiles_to_inchikey` positive test (mock this URL)
- [ ] `_pubchem_smiles_to_inchikey` negative test (mock this URL)
- [ ] `chemical_identifier_resolver` with a correct input, eg: 'aspirin' which should return 'CC(=O)Oc1ccccc1C(O)=O'. (We will mock the URL with status code as 200 and response as given here)
- [ ] `chemical_identifier_resolver` with incorrect input, eg: 'someincorrectinput' which should return a 404 response (This will be mocked)
- [ ] `chemical_identifier_resolver` with incorrect inputs (\n, \t, NoneType, UNPROCESSABLE_INPUT)
- [ ] `encode` with an invalid SMILES input when Chem is not None
- [ ] `encode` with a valid SMILES input when Chem is not None
- [ ] `encode` with incorrect inputs (\n, \t, NoneType, UNPROCESSABLE_INPUT)
- [ ] `encode` with Chem being None with valid smiles and mocked response from `self._pubchem_smiles_to_inchikey(smiles)`
- [ ] `encode` with Chem being None with valid smiles and mocked response from `self._nci_smiles_to_inchikey(smiles)`

2. Tests for `CatalogTable` class in `ersilia/hub/content/catalog.py` file. In this we want to set up a global using a CatalogTable object catalog with some data and columns:
- [ ] `as_list_of_dicts`
- [ ] `as_json`
- [ ] `generate_separator_line`
- [ ] `as_table`
- [ ] `write`

**Comment by musasizivictoria** [comment]
@DhanshreeA let me give this a shot 
thanks

**Comment by DhanshreeA** [comment]
@musasizivictoria! Go for it! Please create a PR and link here.

**Comment by KimFarida** [comment]
May I have the second one please ?

**Comment by musasizivictoria** [comment]
> @musasizivictoria! Go for it! Please create a PR and link here.
PR here:
https://github.com/ersilia-os/ersilia/pull/1320

**Comment by mercybassey** [comment]
Hi @DhanshreeA Can I work on the third one `is_key_header positive test`?

**Comment by DhanshreeA** [comment]
@aderemi1224 would you like to tackle any of the tests here?

**Comment by Ajoke23** [comment]
Hi @DhanshreeA can I pick any of this test to tackle?

**Comment by aderemi1224** [comment]
@

> @aderemi1224 would you like to tackle any of the tests here?

@DhanshreeA  Yes I will, Please assign this to me.
Thank you

**Comment by DhanshreeA** [comment]
@Ajoke23 go ahead! Please link your PR here.

**Comment by DhanshreeA** [comment]
@aderemi1224 you can work on any of the available tests! Thank you.

**Comment by DhanshreeA** [comment]
@KimFarida I had reacted to your comment to indicate yes you can work on any test you wish to take up. Please create a PR and link it here. Thank you!

**Comment by DhanshreeA** [comment]
@musasizivictoria feel free to pick up other tests if you're interested. :)

**Comment by daud99** [comment]
@DhanshreeA  I have already made completed my previous two tasks?

Can I have these tests assigned?

Tests for CatalogTable class in ersilia/hub/content/catalog.py file. In this we want to set up a global using a CatalogTable object catalog with some data and columns:

 as_list_of_dicts
 as_json
 generate_separator_line
 as_table
 write

**Comment by KimFarida** [comment]
> @DhanshreeA I have already made completed my previous two tasks?
> 
> Can I have these tests assigned?
> 
> Tests for CatalogTable class in ersilia/hub/content/catalog.py file. In this we want to set up a global using a CatalogTable object catalog with some data and columns:
> 
> as_list_of_dicts as_json generate_separator_line as_table write

Haha i was working on those 😬

**Comment by daud99** [comment]
LOL @KimFarida I just made the pull request my bad.

**Comment by KimFarida** [comment]
@DhanshreeA i'd asked for the  second one earlier but seeing as @daud99 has worked on them, on to the  others not taken?😬

**Comment by KimFarida** [comment]
> LOL @KimFarida I just made the pull request my bad.

All good 😊. I guess i'll just wait on @DhanshreeA for my next course of action

**Comment by KimFarida** [comment]
> LOL @KimFarida I just made the pull request my bad.

Quick one, did you write all 4 tests or just one?

**Comment by daud99** [comment]
@KimFarida, I believe that instead of working on all four at once, it might be better to tackle them one by one. You could make a pull request for as_json, and for the one you’re working on, try to be explicit with the name instead of using pointers like "second" or "first." This way, it will be easier for others to understand what you're working on and reduce any confusion.

**Comment by KimFarida** [comment]
> @KimFarida, I believe that instead of working on all four at once, it might be better to tackle them one by one. You could make a pull request for as_json, and for the one you’re working on, try to be explicit with the name instead of using pointers like "second" or "first." This way, it will be easier for others to understand what you're working on and reduce any confusion.

Yeah i figured that was my error, my bad.
I did clarify on slack and she said to go ahead🤗.
Thanks !

**Comment by adebisi4145** [comment]
_is_inchikey positive test
PR: #1333

**Comment by adebisi4145** [comment]
Add _is_inchikey negative test
PR: #1334

**Comment by Faith-Adegoke** [comment]
Hi @DhanshreeA. I worked on the `_is_smiles` positive test when Chem is None. Pull Request is #1337.

**Comment by tongyu0924** [comment]
Add test for nci_smiles_to_inchikey and positive inchikey. Pull Request is #1336 #1338

**Comment by musasizivictoria** [comment]
@DhanshreeA 
While looking at  the _is_smiles tests(sppecifically negative test), I noticed that the method currently returns True for empty string inputs. Shouldn’t [_is_smiles](https://github.com/ersilia-os/ersilia/blob/1d5c8f6097d3360fd41783e9db3e4b72c699e146/ersilia/utils/identifiers/compound.py#L42) explicitly reject such cases as invalid
`
if not isinstance(text, str) or not text.strip():
        return False
`
, similar to how non-string inputs are typically handled? Would adding a stricter input validation—like checking for empty or whitespace-only strings—align better with the intended behavior of the method?

otherwise, currently,  testing the  method returns True for an empty string.

cc @Malikbadmus

**Comment by Ajoke23** [comment]
> Hi @DhanshreeA. I worked on the `_is_smiles` positive test when Chem is None. Pull Request is #1337.

Hi @Faith-Adegoke . I already worked on this previously

**Comment by Faith-Adegoke** [comment]
> > Hi @DhanshreeA. I worked on the `_is_smiles` positive test when Chem is None. Pull Request is #1337.
> 
> Hi @Faith-Adegoke . I already worked on this previously #1332

Oh!. Sorry about that. I didn't see it.

**Comment by GemmaTuron** [comment]
Hi @DhanshreeA 

I see this as " In Progress" with no one assigned to it. I also marked it as priority 3. Can we better define this? If it is higher priority we should push for someone to tackle it, if it is low priority maybe put it on hold?

**Comment by DhanshreeA** [comment]
We have made a lot of headway in implementing unit testing within Ersilia thanks to contributions from several people. I will make a small update here as to what is done and what is remaining, sometime this week. Until then, this is good to go back to being on hold.

**Comment by DhanshreeA** [comment]
Hi all the issue is now updated, and ready to be picked up by contributors! Please let us know what you are working on so other contributors can pick up something else.

**Comment by Damisicode** [comment]
@DhanshreeA can i work on the second part of the tests. I mean the tests for the Catalog Table class since they're are all going to be together

**Comment by musasizivictoria** [comment]
@DhanshreeA 

> _pubchem_smiles_to_inchikey negative test (mock this URL)

this test seem already exists, you can mark it done

**Comment by GemmaTuron** [comment]
So this issue will be deprecated. We are refactoring the transformations from smiles to inchikeys etc and the tests will also be revised at that point - let's make sure no legacy code not in use is not lingering at Ersilia @miquelduranfrigola and @Abellegese.

This is linked to: #1584 

I will keep this issue open until we are sure the non used tests are eliminated and the used tests are kept there.

**Comment by Abellegese** [comment]
@GemmaTuron @miquelduranfrigola I think most of the unittests mentioned here are now useless more or less, as we now moved almost entirely using `mdf5` based inchkeys. We better close this and create a better structured issue for other unittest for useful io functions in ersilia.

**Comment by GemmaTuron** [comment]
ok I think it makes sense to clsoe this issue, but before closing it could you spend some time today to create this new issue for future contributors? Thx

**Comment by GemmaTuron** [comment]
@Abellegese kind reminder when you have a bit of time to preparing the issue for future contributors (not actually solving it, just preparing the tasks) 
Thanks :)
